File: sites/_shared/scripts/check-site.mjs
import { readFile, readdir, stat } from 'node:fs/promises';
import { join, resolve, extname } from 'node:path';
import { pathToFileURL } from 'node:url';

const siteDir = process.argv[2];
if (!siteDir) {
  console.error('Usage: node sites/_shared/scripts/check-site.mjs <site-dir>');
  process.exit(1);
}

const root = resolve(siteDir);
const dist = join(root, 'dist');
const { site } = await import(pathToFileURL(join(root, 'site.config.mjs')));
const { tools } = await import(pathToFileURL(join(root, 'data/tools.mjs')));
const { pages } = await import(pathToFileURL(join(root, 'data/pages.mjs')));
const expectedRoutes = ['/', '/tools/', ...tools.map((tool) => `/${tool.slug}/`), ...pages.map((page) => `/${page.slug}/`), '/404.html'];
const compliance = ['/about/', '/contact/', '/privacy-policy/', '/cookie-policy/', '/terms-of-use/', '/disclaimer/'];
const errors = [];
const htmlFiles = [];

async function exists(path) {
  try { await stat(path); return true; } catch { return false; }
}
function routeFile(route) {
  if (route === '/') return join(dist, 'index.html');
  if (route.endsWith('.html')) return join(dist, route);
  return join(dist, route, 'index.html');
}
async function walk(dir) {
  if (!(await exists(dir))) return;
  for (const entry of await readdir(dir)) {
    const path = join(dir, entry);
    const info = await stat(path);
    if (info.isDirectory()) await walk(path);
    else if (extname(path) === '.html') htmlFiles.push(path);
  }
}

for (const route of expectedRoutes) if (!(await exists(routeFile(route)))) errors.push(`Missing route: ${route}`);
for (const route of compliance) if (!expectedRoutes.includes(route)) errors.push(`Required compliance route missing from data: ${route}`);
if (!(await exists(join(dist, 'sitemap.xml')))) errors.push('Missing sitemap.xml');
if (!(await exists(join(dist, 'robots.txt')))) errors.push('Missing robots.txt');
await walk(dist);

const seenTitles = new Map();
const seenDescriptions = new Map();
for (const file of htmlFiles) {
  const html = await readFile(file, 'utf8');
  const route = file.replace(dist, '').replace(/\/index\.html$/, '/').replace(/\/404\.html$/, '/404.html') || '/';
  const title = html.match(/<title>(.*?)<\/title>/)?.[1];
  const desc = html.match(/<meta name="description" content="(.*?)">/)?.[1];
  if (!title) errors.push(`${file}: missing title`);
  else if (title.length < 10) errors.push(`${file}: short title`);
  else if (seenTitles.has(title)) errors.push(`${file}: duplicate title with ${seenTitles.get(title)}`);
  else seenTitles.set(title, file);
  if (!desc || desc.length < 40) errors.push(`${file}: missing/short meta description`);
  else if (seenDescriptions.has(desc)) errors.push(`${file}: duplicate meta description with ${seenDescriptions.get(desc)}`);
  else seenDescriptions.set(desc, file);
  for (const needle of ['rel="canonical"', 'property="og:title"', 'name="twitter:card"']) if (!html.includes(needle)) errors.push(`${file}: missing ${needle}`);
  const h1Count = (html.match(/<h1[\s>]/g) || []).length;
  if (h1Count !== 1) errors.push(`${file}: expected one h1, found ${h1Count}`);
  if (/TODO|Lorem ipsum|Insert Date|placeholder content/i.test(html)) errors.push(`${file}: contains placeholder/TODO content`);
  for (const link of [...html.matchAll(/href="(\/[^"#?]*)/g)].map((match) => match[1])) {
    if (link.startsWith('/assets/') || ['/sitemap.xml', '/robots.txt'].includes(link)) continue;
    const normalized = link.endsWith('/') || link.endsWith('.html') ? link : `${link}/`;
    if (!expectedRoutes.includes(normalized)) errors.push(`${file}: broken internal link ${link}`);
  }
  if (tools.some((tool) => route === `/${tool.slug}/`)) {
    for (const needle of ['FAQPage', 'SoftwareApplication', 'data-reset-tool', 'Copy Result', 'Formula', 'Related tools', 'Advertisement']) {
      if (!html.includes(needle)) errors.push(`${file}: tool page missing ${needle}`);
    }
    const relatedLinks = (html.match(/class="related"[\s\S]*?<\/div>/)?.[0].match(/<a /g) || []).length;
    if (relatedLinks < 5) errors.push(`${file}: expected at least 5 related links, found ${relatedLinks}`);
  }
}

const sitemap = await readFile(join(dist, 'sitemap.xml'), 'utf8').catch(() => '');
const robots = await readFile(join(dist, 'robots.txt'), 'utf8').catch(() => '');
for (const route of expectedRoutes.filter((route) => route !== '/404.html')) {
  if (!sitemap.includes(route === '/' ? `${site.baseUrl}/` : `${site.baseUrl}${route}`)) errors.push(`sitemap missing ${route}`);
}
if (!robots.includes(`${site.baseUrl}/sitemap.xml`)) errors.push('robots missing correct sitemap URL');

if (errors.length) {
  console.error(errors.join('\n'));
  process.exit(1);
}
console.log(`${site.name} check passed: ${expectedRoutes.length} expected routes, ${htmlFiles.length} HTML files.`);

File: sites/_shared/scripts/check_site.py
from pathlib import Path
import re
import sys

if len(sys.argv) < 2:
    print('Usage: python3 sites/_shared/scripts/check_site.py <site-dir>', file=sys.stderr)
    raise SystemExit(1)

root = Path(sys.argv[1])
key = root.name
base_urls = {
    'datecalctools': 'https://datecalctools.com',
    'unitconverthub': 'https://unitconverthub.com',
    'quickmathtools': 'https://quickmathtools.com',
    'wagemoneycalc': 'https://wagemoneycalc.com',
    'healthlifecalc': 'https://healthlifecalc.com',
}
tool_routes = {
    'datecalctools': ['age-calculator','date-calculator','days-between-dates-calculator','business-days-calculator','time-duration-calculator','countdown-calculator','birthday-calculator','weekday-calculator','weeks-between-dates-calculator','months-between-dates-calculator'],
    'unitconverthub': ['cm-to-inches','inches-to-cm','kg-to-lbs','lbs-to-kg','celsius-to-fahrenheit','fahrenheit-to-celsius','meters-to-feet','feet-to-meters','ml-to-oz','oz-to-ml','liters-to-gallons','gallons-to-liters','square-feet-to-square-meters','km-to-miles','miles-to-km'],
    'quickmathtools': ['percentage-calculator','percentage-increase-calculator','percentage-decrease-calculator','fraction-calculator','fraction-to-decimal-calculator','average-calculator','ratio-calculator','random-number-generator','grade-calculator','gpa-calculator'],
    'wagemoneycalc': ['hourly-wage-calculator','salary-calculator','salary-to-hourly-calculator','overtime-calculator','discount-calculator','sales-tax-calculator','tip-calculator','profit-margin-calculator','freelance-rate-calculator','unit-price-calculator'],
    'healthlifecalc': ['bmi-calculator','bmr-calculator','calorie-calculator','water-intake-calculator','ideal-weight-calculator','body-fat-calculator','pregnancy-due-date-calculator','ovulation-calculator','macro-calculator','steps-to-calories-calculator'],
}
base = base_urls[key]
expected = ['/','/tools/'] + [f'/{slug}/' for slug in tool_routes[key]] + ['/about/','/contact/','/privacy-policy/','/cookie-policy/','/terms-of-use/','/disclaimer/','/404.html']
dist = root / 'dist'
errors = []

def file_for(route):
    if route == '/': return dist / 'index.html'
    if route.endswith('.html'): return dist / route.lstrip('/')
    return dist / route.lstrip('/') / 'index.html'

for route in expected:
    if not file_for(route).exists(): errors.append(f'Missing route: {route}')
for name in ['sitemap.xml','robots.txt','assets/css/styles.css','assets/js/runtime.js','assets/js/site-tools.js']:
    if not (dist / name).exists(): errors.append(f'Missing {name}')

html_files = list(dist.rglob('*.html')) if dist.exists() else []
titles = {}
descs = {}
for file in html_files:
    html = file.read_text(encoding='utf-8')
    route = str(file.relative_to(dist)).replace('index.html','')
    route = '/' + route if not route.startswith('/') else route
    if route == '/404.html': pass
    title = re.search(r'<title>(.*?)</title>', html)
    desc = re.search(r'<meta name="description" content="(.*?)">', html)
    if not title: errors.append(f'{file}: missing title')
    elif title.group(1) in titles: errors.append(f'{file}: duplicate title with {titles[title.group(1)]}')
    else: titles[title.group(1)] = file
    if not desc or len(desc.group(1)) < 40: errors.append(f'{file}: missing/short meta description')
    elif desc.group(1) in descs: errors.append(f'{file}: duplicate description with {descs[desc.group(1)]}')
    else: descs[desc.group(1)] = file
    for needle in ['rel="canonical"','property="og:title"','name="twitter:card"']:
        if needle not in html: errors.append(f'{file}: missing {needle}')
    h1s = len(re.findall(r'<h1[\s>]', html))
    if h1s != 1: errors.append(f'{file}: expected one h1, found {h1s}')
    if re.search(r'TODO|Lorem ipsum|Insert Date|placeholder content', html, re.I):
        errors.append(f'{file}: contains placeholder/TODO content')
    for link in re.findall(r'href="(/[^"#?]*)', html):
        if link.startswith('/assets/') or link in ['/sitemap.xml','/robots.txt']: continue
        normalized = link if link.endswith('/') or link.endswith('.html') else link + '/'
        if normalized not in expected: errors.append(f'{file}: broken internal link {link}')
    if route in [f'/{slug}/' for slug in tool_routes[key]]:
        for needle in ['FAQPage','SoftwareApplication','data-reset-tool','Copy Result','Formula','Related tools','Advertisement']:
            if needle not in html: errors.append(f'{file}: tool page missing {needle}')
        related = re.search(r'class="related"[\s\S]*?</div>', html)
        count = len(re.findall(r'<a ', related.group(0))) if related else 0
        if count < 5: errors.append(f'{file}: expected at least 5 related links, found {count}')
        if key == 'wagemoneycalc' and 'financial, tax, legal, or payroll advice' not in html:
            errors.append(f'{file}: missing finance disclaimer')
        if key == 'healthlifecalc' and 'not medical advice, diagnosis, or treatment' not in html:
            errors.append(f'{file}: missing health disclaimer')

sitemap = (dist / 'sitemap.xml').read_text(encoding='utf-8') if (dist / 'sitemap.xml').exists() else ''
robots = (dist / 'robots.txt').read_text(encoding='utf-8') if (dist / 'robots.txt').exists() else ''
for route in expected:
    if route == '/404.html': continue
    url = base + ('/' if route == '/' else route)
    if url not in sitemap: errors.append(f'sitemap missing {route}')
if f'{base}/sitemap.xml' not in robots: errors.append('robots missing correct sitemap URL')

if errors:
    print('\n'.join(errors))
    raise SystemExit(1)
print(f'{key} check passed: {len(expected)} expected routes, {len(html_files)} HTML files.')
